images/kindnetd/cmd/kindnetd/subnets.go

```diff
@@ -0,0 +1,102 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+	"regexp"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/yaml"
+)
+
+type subnets struct {
+	pods     sets.String
+	services sets.String
+}
+
+func newSubnets() *subnets {
+	return &subnets{
+		pods:     sets.NewString(),
+		services: sets.NewString(),
+	}
+}
+
+func (s *subnets) All() []string {
+	return s.pods.Union(s.services).List()
+}
+
+func getNoMasqueradeSubnets(clientset *kubernetes.Clientset) []string {
+	s := newSubnets()
+
+	// try getting from kubeadm config
+	kubeadmSubnets := getKubeadmSubnets(clientset)
+	s.pods = s.pods.Union(kubeadmSubnets.pods)
+	s.services = s.services.Union(kubeadmSubnets.services)
+
+	// TODO: we have other fallback options for completeness including:
+	// - kube-apiserver flags
+	// - component config
+	// - defaults
+	return s.All()
+}
+
+const (
+	kubeadmServiceCIDRRegex = `(?m)^\s*serviceSubnet: (.*)\n`
```

Comment:
@neolit123 are these fields `serviceSubnet` and `podSubnet` stable in kubeadm, are they following the API policies? also, you know if there were changes? if there were changes we should keep the previous environment variable configuration for backwards compatibility

Reply:
Kubeadm config is beta.
Even if they weren't stae, we'd just make a new version of kindnetd when they change.

> if there were changes we should keep the previous environment variable configuration for backwards compatibility

This is for comparability with the kind binary only, when templating. If we are in fact only doing podSubnet it makes more sense, but even then we don't need it.

Reply:
> also, you know if there were changes?

kubeadm has been exposing these fields since at least 1.9 / early-v1alpha1.